datafusion/physical-plan/src/joins/cross_join.rs

```diff
@@ -374,64 +346,130 @@ impl Stream for CrossJoinStream {
 }
 
 impl CrossJoinStream {
-    /// Separate implementation function that unpins the [`CrossJoinStream`] so
-    /// that partial borrows work correctly
+    /// Separate implementation function that unpins the [`CrossJoinStream`]
+    /// so that partial borrows work correctly
     fn poll_next_impl(
         &mut self,
         cx: &mut std::task::Context<'_>,
     ) -> std::task::Poll<Option<Result<RecordBatch>>> {
+        let schema = self.schema();
+
         let build_timer = self.join_metrics.build_time.timer();
         let (left_data, _) = match ready!(self.left_fut.get(cx)) {
```

Comment:
```rust
let (left_data, _) = match ready!(self.left_fut.get(cx)) {
    Ok(left_data) => left_data,
    Err(e) => return Poll::Ready(Some(Err(e))),
};
``` 
this shouldn't need to be polled continuously, since there is no need for that. Just clone it and write to your state, then use it. Vec<RecordBatch> is quite simple to clone.